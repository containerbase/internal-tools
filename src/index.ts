// Node v16 is missing structuredClone, so we polyfill it here.
import 'core-js/actual/structured-clone';
import run from './runner';

run().catch(console.error);
