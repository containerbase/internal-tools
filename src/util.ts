import { exec as _exec } from '@actions/exec';
import { ExecOptions as _ExecOptions } from '@actions/exec/lib/interfaces';
import log from './utils/logger';
import { getInput, startGroup, endGroup } from '@actions/core';
import { join } from 'path';
import fs from 'fs';
import { promisify } from 'util';
import findUp = require('find-up');

const readFileAsync = promisify(fs.readFile);

export type ExecOptions = _ExecOptions;

export type ExecResult = {
  readonly code: number;
  readonly stdout: string;
  readonly stderr: string;
};

export async function exec(
  cmd: string,
  args: string[],
  options?: ExecOptions
): Promise<ExecResult> {
  let stdout = '';
  let stderr = '';
  let code: number;

  try {
    startGroup(`${cmd} ${args.join(' ')}`);
    code = await _exec(cmd, args, {
      ...options,
      listeners: {
        stdout: (data: Buffer) => {
          stdout += data.toString();
        },
        stderr: (data: Buffer) => {
          stderr += data.toString();
        },
      },
    });
  } finally {
    endGroup();
  }
  if (code) {
    log.error({ code });
    throw new Error('Command failed');
  }

  return { code, stdout, stderr };
}

export function isDryRun(): boolean {
  const val = getInput('dry-run');
  return !!val && val === 'true';
}

/**
 * Get environment variable or empty string.
 * Used for easy mocking.
 * @param key variable name
 */
export function getEnv(key: string): string {
  return process.env[key] ?? '';
}

export async function readJson<T = unknown>(file: string): Promise<T> {
  const path = join(getEnv('GITHUB_WORKSPACE'), file);
  const res = await import(path);
  // istanbul ignore next
  return res?.default ?? res;
}

export async function readFile(file: string): Promise<string> {
  const path = join(getEnv('GITHUB_WORKSPACE'), file);
  return await readFileAsync(path, 'utf8');
}

export const MultiArgsSplitRe = /\s*(?:[;,]|$)\s*/;

export function getArg(name: string, opts?: { required?: boolean }): string;
export function getArg(
  name: string,
  opts?: { required?: boolean; multi: true }
): string[];
export function getArg(
  name: string,
  opts?: { required?: boolean; multi?: boolean }
): string | string[];
export function getArg(
  name: string,
  opts?: { required?: boolean; multi?: boolean }
): string | string[] {
  const val = getInput(name, opts);
  return opts?.multi ? val.split(MultiArgsSplitRe).filter(Boolean) : val;
}

let _pkg: Promise<string | undefined>;

/**
 * Resolve path for a file relative to renovate root directory (our package.json)
 * @param file a file to resolve
 */
export async function resolveFile(file: string): Promise<string> {
  if (!_pkg) {
    _pkg = findUp('package.json', { cwd: __dirname, type: 'file' });
  }
  const pkg = await _pkg;
  // istanbul ignore if
  if (!pkg) {
    throw new Error('Missing package.json');
  }
  return join(pkg, '../', file);
}
