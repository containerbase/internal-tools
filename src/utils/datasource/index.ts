import { getDatasources } from 'renovate/dist/modules/datasource';
import log from '../logger';

export function register(): void {
  log('register datasources');
  getDatasources();
}
