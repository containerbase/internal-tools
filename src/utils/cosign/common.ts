import { exec } from '../../util';
import type { ExecResult } from '../types';

export async function cosign(...args: string[]): Promise<ExecResult> {
  process.env['COSIGN_EXPERIMENTAL'] = 'true';
  return await exec('cosign', [...args]);
}
