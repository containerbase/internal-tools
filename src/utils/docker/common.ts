import { exec } from '../../util';
import log from '../logger';
import type { ExecResult } from '../types';

export enum DockerPlatform {
  x86_64 = 'linux/amd64',
  aarch64 = 'linux/arm64',
}

export type DockerArch = 'x86_64' | 'aarch64';

export async function docker(...args: string[]): Promise<ExecResult> {
  return await exec('docker', [...args]);
}
export async function dockerRun(...args: string[]): Promise<void> {
  await docker('run', '--rm', ...args);
}

export async function dockerBuildx(...args: string[]): Promise<ExecResult> {
  return await docker('buildx', ...args);
}

type DockerTagConfig = {
  image: string;
  imagePrefix: string;
  src: string;
  tgt: string;
};

export async function dockerTag({
  image,
  imagePrefix,
  src,
  tgt,
}: DockerTagConfig): Promise<ExecResult> {
  return await exec('docker', [
    'tag',
    `${imagePrefix}/${image}:${src}`,
    `${imagePrefix}/${image}:${tgt}`,
  ]);
}

export async function dockerPrune(): Promise<void> {
  log('Pruning docker system');
  await docker('system', 'prune', '--force', '--all');
}

export async function dockerDf(): Promise<void> {
  log('Docker system disk usage');
  await docker('system', 'df');
}
