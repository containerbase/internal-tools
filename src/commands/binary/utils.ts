import { getInput } from '@actions/core';
import is from '@sindresorhus/is';
import { getDefaultVersioning } from 'renovate/dist/modules/datasource';
import { getArch, getArg, getDistro, isDryRun, readJson } from '../../util';
import { readDockerConfig } from '../../utils/config';
import {
  DockerPlatform,
  dockerBuildx,
  dockerRun,
} from '../../utils/docker/common';
import log from '../../utils/logger';
import type { BinaryBuilderConfig, ConfigFile } from '../../utils/types';

export async function getConfig(): Promise<BinaryBuilderConfig> {
  const configFile = getInput('config') || 'builder.json';

  const cfg = await readJson<ConfigFile>(configFile);

  if (!is.object(cfg)) {
    throw new Error('missing-config');
  }

  if (!is.string(cfg.image)) {
    cfg.image = getInput('image', { required: true });
  }

  if (!is.string(cfg.buildArg)) {
    cfg.buildArg = cfg.image.toUpperCase() + '_VERSION';
  }

  await readDockerConfig(cfg);

  return {
    ...cfg,
    ignoredVersions: cfg.ignoredVersions ?? [],
    dryRun: isDryRun(),
    lastOnly: getArg('last-only') == 'true',
    buildArgs: getArg('build-args', { multi: true }),
    versioning: cfg.versioning ?? getDefaultVersioning(cfg.datasource),
  } as BinaryBuilderConfig;
}

export async function createBuilderImage(
  ws: string,
  { buildArgs }: BinaryBuilderConfig
): Promise<void> {
  log('Creating builder image');
  const args = [
    'build',
    '--load',
    '-t',
    'builder',
    '--build-arg',
    `DISTRO=${getDistro()}`,
  ];
  const arch = getArch();
  if (is.nonEmptyString(arch)) {
    args.push('--platform', DockerPlatform[arch]);
  }
  if (is.nonEmptyArray(buildArgs)) {
    args.push(...buildArgs.map((b) => `--build-arg=${b}`));
  }
  await dockerBuildx(...args, ws);
}

export async function runBuilder(ws: string, version: string): Promise<void> {
  const args = ['--name', 'builder', '--volume', `${ws}/.cache:/cache`];
  const arch = getArch();
  if (is.nonEmptyString(arch)) {
    args.push('--platform', DockerPlatform[arch]);
  }
  await dockerRun(...args, 'builder', version);
}
