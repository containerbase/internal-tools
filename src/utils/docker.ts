import is from '@sindresorhus/is';
import * as chalk from 'chalk';
import * as delay from 'delay';
import got, { HTTPError, Headers } from 'got';
import * as wwwAuthenticate from 'www-authenticate';
import { exists } from '../util';
import { cosign } from './cosign/common';
import { docker } from './docker/common';
import log from './logger';
import { ExecError } from './types';

export function splitImage(image: string): {
  registry: string;
  repository: string;
} {
  const parts = image.split('/');
  if (parts.length > 2) {
    return { registry: parts[0], repository: parts.slice(1).join('/') };
  } else {
    return { registry: 'index.docker.io', repository: image };
  }
}

export type DockerManifestV2 = {
  schemaVersion: number;
  config: {
    digest: string;
  };
};

export async function getAuthHeaders(
  registry: string,
  repository: string
): Promise<Headers> {
  try {
    const apiCheckUrl = `https://${registry}/v2/`;
    const apiCheckResponse = await got(apiCheckUrl, { throwHttpErrors: false });
    if (apiCheckResponse.headers['www-authenticate'] === undefined) {
      return {};
    }
    const authenticateHeader = new wwwAuthenticate.parsers.WWW_Authenticate(
      apiCheckResponse.headers['www-authenticate']
    );

    const authUrl = `${authenticateHeader.parms.realm}?service=${authenticateHeader.parms.service}&scope=repository:${repository}:pull`;
    const authResponse = (
      await got<{ token?: string; access_token?: string }>(authUrl, {
        responseType: 'json',
      })
    ).body;

    const token = authResponse.token || authResponse.access_token;
    if (!token) {
      throw new Error('Failed to obtain docker registry token');
    }
    return {
      authorization: `Bearer ${token}`,
    };
  } catch (err) {
    log.error(chalk.red('auth error'), (err as Error).message);
    throw new Error('Failed to obtain docker registry token');
  }
}

export enum DockerContentType {
  ManifestV1 = 'application/vnd.docker.distribution.manifest.v1+json',
  ManifestV1Signed = 'application/vnd.docker.distribution.manifest.v1+prettyjws',
  ManifestV2 = 'application/vnd.docker.distribution.manifest.v2+json',
}

const shaRe = /(sha256:[a-f0-9]{64})/;

export async function getRemoteImageId(
  registry: string,
  repository: string,
  tag = 'latest'
): Promise<string> {
  const headers = await getAuthHeaders(registry, repository);
  headers.accept = DockerContentType.ManifestV2;
  const url = `https://${registry}/v2/${repository}/manifests/${tag}`;

  try {
    const resp = await got<DockerManifestV2>(url, {
      headers,
      responseType: 'json',
    });

    switch (resp.headers['content-type']) {
      case DockerContentType.ManifestV2:
        return resp.body.config.digest;

      case DockerContentType.ManifestV1:
      case DockerContentType.ManifestV1Signed:
        // something wrong, we need to overwrite existing
        log.warn(
          chalk.yellow('Wrong response'),
          `Wrong response: ${resp.headers['content-type'] as string}`
        );
        return '<error>';

      default:
        throw new Error(
          `Unsupported response: ${resp.headers['content-type'] as string}`
        );
    }
  } catch (e) {
    if (e instanceof HTTPError && e.response.statusCode === 404) {
      // no image published yet
      return '<none>';
    }
    log.error(chalk.red('request error'), (e as Error).message);
    throw new Error('Could not find remote image id');
  }
}

export async function getLocalImageId(
  image: string,
  tag = 'latest'
): Promise<string> {
  const res = await docker('inspect', "--format='{{.Id}}'", `${image}:${tag}`);

  const [, id] = shaRe.exec(res.stdout) ?? [];

  if (!id) {
    log.error(res);
    throw new Error('Could not find local image id');
  }

  return id;
}

export type BuildOptions = {
  image: string;
  imagePrefix: string;
  cache?: string;
  cacheTags?: string[];
  tag?: string;
  dryRun?: boolean;
  buildArgs?: string[];
};

const errors = [
  'unexpected status: 400 Bad Request',
  ': no response',
  'error writing layer blob',
];

function canRetry(err: ExecError): boolean {
  return errors.some((str) => err.stderr.includes(str));
}

export async function build({
  image,
  imagePrefix,
  cache,
  cacheTags,
  tag = 'latest',
  dryRun,
  buildArgs,
}: BuildOptions): Promise<void> {
  const args = [
    'buildx',
    'build',
    '--load',
    `--tag=${imagePrefix}/${image}:${tag}`,
  ];

  if (is.nonEmptyArray(buildArgs)) {
    args.push(...buildArgs.map((b) => `--build-arg=${b}`));
  }

  if (is.string(cache)) {
    const cacheImage = `${imagePrefix}/${cache}:${image.replace(/\//g, '-')}`;
    args.push(`--cache-from=${cacheImage}-${tag}`);

    if (is.nonEmptyArray(cacheTags)) {
      for (const ctag of cacheTags) {
        args.push(`--cache-from=${cacheImage}-${ctag}`);
      }
    }

    if (!dryRun) {
      args.push(`--cache-to=type=registry,ref=${cacheImage}-${tag},mode=max`);
    }
  }

  for (let build = 0; ; build++) {
    try {
      await docker(...args, '.');
      break;
    } catch (e) {
      if (e instanceof ExecError && canRetry(e) && build < 2) {
        log.error(chalk.red(`docker build error on try ${build}`), e);
        await delay(5000);
        continue;
      }
      throw e;
    }
  }
}

type PublishOptions = {
  image: string;
  imagePrefix: string;
  tag: string;
  dryRun?: boolean;
};

export async function publish({
  image,
  imagePrefix,
  tag,
  dryRun,
}: PublishOptions): Promise<void> {
  const imageName = `${imagePrefix}/${image}`;
  const fullName = `${imageName}:${tag}`;
  log.info(chalk.blue('Processing image:'), chalk.yellow(fullName));

  log('Fetch new id');
  const newId = await getLocalImageId(imageName, tag);

  log('Fetch old id');
  const { registry, repository } = splitImage(imageName);
  const oldId = await getRemoteImageId(registry, repository, tag);

  if (oldId === newId) {
    log('Image uptodate, no push nessessary:', chalk.yellow(oldId));
    return;
  }

  log('Publish new image', `${oldId} => ${newId}`);
  if (dryRun) {
    log.warn(chalk.yellow('[DRY_RUN]'), chalk.blue('Would push:'), fullName);
  } else {
    await docker('push', fullName);
    await sign(fullName);
  }
  log.info(chalk.blue('Processing image finished:', newId));
}

async function sign(fullName: string): Promise<void> {
  if (!(await exists('cosign'))) {
    log.warn('Cosign is not installed. Skipping container signing');
    return;
  }

  log('Signing image', fullName);
  await cosign('sign', fullName);
}
