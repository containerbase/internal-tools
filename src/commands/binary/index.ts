import 'source-map-support/register';
import { setFailed } from '@actions/core';
import * as chalk from 'chalk';
import * as shell from 'shelljs';
import { getArg, getWorkspace } from '../../util';
import { addHostRule, getBuildList } from '../../utils/builds';
import { init } from '../../utils/docker/buildx';
import {
  getOctokit,
  hasAsset,
  updateRelease,
  uploadAsset,
} from '../../utils/github';
import log from '../../utils/logger';
import { createBuilderImage, getConfig, runBuilder } from './utils';

let toBuild = 99;

export async function run(): Promise<void> {
  try {
    log.info('Builder started');
    const ws = getWorkspace();

    const cfg = await getConfig();

    if (cfg.dryRun) {
      log.warn(chalk.yellow('[DRY_RUN] detected'));
      cfg.lastOnly = true;
    }

    const token = getArg('token', { required: true });
    const api = getOctokit(token);

    addHostRule({ matchHost: 'github.com', token });

    log('config:', JSON.stringify(cfg));

    const builds = await getBuildList(cfg);

    if (!builds?.versions.length) {
      setFailed(`No versions found.`);
      return;
    }

    shell.mkdir('-p', `${ws}/.cache`);

    await init();

    await createBuilderImage(ws, cfg);

    const failed: string[] = [];

    for (const version of builds.versions) {
      await updateRelease(api, cfg, version);
      if (await hasAsset(api, cfg, version)) {
        if (cfg.dryRun) {
          log.warn(
            chalk.yellow('[DRY_RUN] Would skipp existing version:'),
            version
          );
        } else {
          log('Skipping existing version:', version);
          continue;
        }
      }

      // istanbul ignore if
      if (toBuild-- <= 0) {
        log.info('Build limit reached');
        break;
      }

      log.info('Processing version:', version);

      try {
        log('Runing builder:', version);
        await runBuilder(ws, version);

        if (cfg.dryRun) {
          log.warn(
            chalk.yellow('[DRY_RUN] Would upload release asset:'),
            version
          );
        } else {
          log('Uploading release:', version);
          await uploadAsset(api, cfg, version);
        }
      } catch (e) {
        failed.push(version);
        // eslint-disable-next-line
        log(`Version ${version} failed: ${e.message}`, e.stack);
      }
    }

    if (failed.length) {
      setFailed(`Versions failed: ${failed.join(', ')}`);
    }
  } catch (error) {
    log((error as Error).stack);
    setFailed(error as Error);
  }
}
