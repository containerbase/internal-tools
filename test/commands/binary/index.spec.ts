/* eslint-disable @typescript-eslint/no-var-requires */
import * as _core from '@actions/core';
import * as _datasources from 'renovate/dist/modules/datasource';
import { run } from '../../../src/commands/binary';
import * as _utils from '../../../src/util';
import * as _docker from '../../../src/utils/docker/common';
import * as _github from '../../../src/utils/github';
import { mocked } from '../../utils';

jest.mock('renovate/dist/modules/datasource');
jest.mock('../../../src/util');
jest.mock('../../../src/utils/docker/common');
jest.mock('../../../src/utils/docker/buildx', () => ({
  init: () => Promise.resolve(),
}));
jest.mock('../../../src/utils/github');
jest.mock('../../../src/utils/datasource');

const core = mocked(_core);
const utils = mocked(_utils);
const docker = mocked(_docker);
const datasources = mocked(_datasources);
const github = mocked(_github);
const version = '2.7.2';

describe('commands/binary/index', () => {
  let input: Record<string, string>;

  beforeEach(() => {
    jest.resetAllMocks();
    input = {};
    core.getInput.mockImplementation((k) => input[k]);
    utils.readJson.mockResolvedValueOnce(require('./__fixtures__/ruby.json'));
    utils.getArg.mockImplementation((_, o) => (o?.multi ? [] : ''));
    utils.getWorkspace.mockReturnValue('.');
  });

  it('works ruby', async () => {
    utils.readFile.mockResolvedValue(
      `# renovate: datasource=ruby-version depName=ruby-version versioning=ruby\nARG RUBY_VERSION=${version}\n`
    );
    datasources.getPkgReleases.mockResolvedValueOnce({
      releases: [{ version }, { version: '3.0.0-rc.24' }],
    });

    github.hasAsset.mockResolvedValueOnce(true);

    await run();

    expect(docker.dockerRun.mock.calls).toMatchSnapshot('docker');
  });

  it('works ruby (dry-run)', async () => {
    utils.isDryRun.mockReturnValueOnce(true);
    utils.readFile.mockResolvedValue(
      `# renovate: datasource=ruby-version depName=ruby-version versioning=ruby\nARG RUBY_VERSION=${version}\n`
    );
    datasources.getPkgReleases.mockResolvedValueOnce({
      releases: [{ version }, { version: '3.0.0-rc.24' }],
    });

    github.hasAsset.mockResolvedValueOnce(true);

    await run();

    expect(docker.dockerRun.mock.calls).toMatchSnapshot('docker');
  });

  it('works ruby (no-assets)', async () => {
    utils.readFile.mockResolvedValue(
      `# renovate: datasource=ruby-version depName=ruby-version versioning=ruby\nARG RUBY_VERSION=${version}\n`
    );
    datasources.getPkgReleases.mockResolvedValueOnce({
      releases: [{ version }, { version: '3.0.0-rc.24' }],
    });

    github.hasAsset.mockResolvedValueOnce(false);

    await run();

    expect(docker.dockerRun.mock.calls).toMatchSnapshot('docker');
  });

  it('works dummy', async () => {
    utils.readJson.mockReset();
    utils.readJson.mockResolvedValueOnce(require('./__fixtures__/dummy.json'));
    utils.readFile.mockResolvedValue(``);

    github.hasAsset.mockResolvedValueOnce(true);

    await run();

    expect(docker.dockerRun.mock.calls).toMatchSnapshot('docker');
    expect(datasources.getPkgReleases).not.toHaveBeenCalled();
  });

  it('works python (dry-run)', async () => {
    utils.isDryRun.mockReturnValueOnce(true);
    utils.readJson.mockReset();
    utils.readJson.mockResolvedValueOnce(require('./__fixtures__/python.json'));
    utils.readFile.mockResolvedValue(
      `# renovate: datasource=docker depName=python versioning=docker\nARG PYTHON_VERSION=3.9.1`
    );

    github.hasAsset.mockResolvedValueOnce(true);

    await run();

    expect(docker.dockerRun.mock.calls).toMatchSnapshot('docker');
    expect(datasources.getPkgReleases).not.toHaveBeenCalled();
  });

  it('empty releases', async () => {
    datasources.getPkgReleases.mockResolvedValueOnce({
      releases: [],
    });

    await run();
    expect(docker.dockerRun).not.toHaveBeenCalled();
    expect(github.updateRelease).not.toHaveBeenCalled();
    expect(core.setFailed).toHaveBeenCalledWith('No versions found.');
  });

  it('null releases', async () => {
    datasources.getPkgReleases.mockResolvedValueOnce(null);
    await run();
    expect(docker.dockerRun).not.toHaveBeenCalled();
    expect(github.updateRelease).not.toHaveBeenCalled();
    expect(core.setFailed).toHaveBeenCalledWith('No versions found.');
  });

  it('catch errors', async () => {
    utils.readFile.mockResolvedValue(
      `# renovate: datasource=ruby-version depName=ruby-version versioning=ruby\nARG RUBY_VERSION=${version}\n`
    );
    datasources.getPkgReleases.mockResolvedValueOnce({
      releases: [{ version }, { version: '3.0.0-rc.24' }],
    });

    github.hasAsset.mockRejectedValueOnce(new Error('dummy'));

    await run();

    expect(core.setFailed).toHaveBeenCalledTimes(1);
    expect(core.setFailed.mock.calls).toMatchSnapshot();
  });

  it('continues on errors', async () => {
    utils.readFile.mockResolvedValue(
      `# renovate: datasource=ruby-version depName=ruby-version versioning=ruby\nARG RUBY_VERSION=${version}\n`
    );
    datasources.getPkgReleases.mockResolvedValueOnce({
      releases: [{ version }, { version: '3.0.0-rc.24' }],
    });

    docker.dockerRun.mockRejectedValueOnce(new Error('dummy'));

    await run();

    expect(core.setFailed).toHaveBeenCalledTimes(1);
    expect(core.setFailed.mock.calls).toMatchSnapshot();
  });
});
