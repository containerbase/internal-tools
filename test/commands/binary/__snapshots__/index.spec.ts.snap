// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`commands/binary/index catch errors 1`] = `
Array [
  Array [
    [Error: dummy],
  ],
]
`;

exports[`commands/binary/index continues on errors 1`] = `
Array [
  Array [
    "Versions failed: 2.7.2",
  ],
]
`;

exports[`commands/binary/index works dummy: docker 1`] = `
Array [
  Array [
    "--name",
    "builder",
    "--volume",
    "./.cache:/cache",
    "builder",
    "13",
  ],
]
`;

exports[`commands/binary/index works python (dry-run): docker 1`] = `
Array [
  Array [
    "--name",
    "builder",
    "--volume",
    "./.cache:/cache",
    "builder",
    "3.1.0-rc",
  ],
]
`;

exports[`commands/binary/index works ruby (dry-run): docker 1`] = `
Array [
  Array [
    "--name",
    "builder",
    "--volume",
    "./.cache:/cache",
    "builder",
    "2.7.2",
  ],
]
`;

exports[`commands/binary/index works ruby (no-assets): docker 1`] = `
Array [
  Array [
    "--name",
    "builder",
    "--volume",
    "./.cache:/cache",
    "builder",
    "2.7.2",
  ],
]
`;

exports[`commands/binary/index works ruby: docker 1`] = `Array []`;
